### ISSUE TITLE
[RFC] Flop counters in PyTorch

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#93121]
### 🚀 The feature, motivation and pitch

Lots of people have built flop counters/asked for FLOP counters (or other performance counters) in PyTorch. There's a litany of requests, but some common ones include: 1. coverage (i.e. no restrictions on how the model is written, capturing backwards, etc.), 2. module-level grouping, 3. customization (i.e. being able to modify flop formulas), and 4. Capturing metrics other than FLOPs (like memory bandwidth).

There's another factor that's relevant with `PyTorch 2.0`, which is whether it works in eager-mode or under `torch.compile`. Some of these features are a lot of easier to do in eager mode (coverage/module-level grouping), and others are a lot easier to do in compiled mode (capturing memory bandwidth or non-matmul FLOPs).

I propose we add 2 different flop counters to PyTorch, one that's eager based and one that's graph based.

The eager one (implemented similarly to [this](https://dev-discuss.pytorch.org/t/the-ideal-pytorch-flop-counter-with-torch-dispatch/505/5)).
- Works with all operations in eager mode (forwards, backwards, checkpointing, functorch, w.e.)
- Only counts GMACs (no memory accesses, no non-matmul FLOPs)
- Provides module-level breakdowns
- Provides a context manager API where it counts all flops executed within the context manager.
- Provides an API where you can override formulas for each (aten) op.

Compiled flop counter:
- Works with ... most operations in eager mode (the stuff that torch.compile can compile at least).
- Can count FLOPs (for some formula for it...) and memory accesses.
- Does not provide module-level breakdowns.
- Probably also a context manager API.
- No easy API for overriding formulas for each op (I guess we can provide a way to say, map cos to however many flops you want, but that's less clear).

cc: @Jokeren @rwightman, @ppwwyyxx. @ezyang

### Alternatives

N/A

**Comment by albanD** [comment]
Thanks for opening this topic Horace!

> I propose we add 2 different flop counters to PyTorch,

I'm sure there are going to be a lot of requests for different features here. I think that if we want to keep this in core, we want a simple and extensible design. But if we want to have a version will all the features and the knobs people want, I feel like a small python-only side repo might be a better place for it. WDYT?
Either direction we take, it should be clearly stated how complex we're ready to make it.

> no non-matmul FLOPs

So for the eager version you really ignore everything that is not a matmul like conv/activations/normalization ? I feel like I'm missing something?

> Only counts GMACs

For the eager version, can we make the design extensible such that if someone wants to add FLOPS it is easy?


Also I'm sure you're going to be interested @anupambhatnagar @fmassa

**Comment by vadimkantorov** [comment]
This has also existed in fvcore: https://github.com/facebookresearch/fvcore/blob/main/docs/flop_count.md

probably used in detectron2 or somewhere else

This is somewhat related to https://github.com/pytorch/vision/issues/7070 where I proposed to integrate useful utils from fvcore/detectron2 into core (maybe into torch.utils-> some underscore namespace)/torchvision because this code is already quite old and wideness of usage can now already be talked about)

**Comment by rwightman** [comment]
No non-matmul FLOPS aka MACs is the core feature IMHO, that's what is quoted the most, even when incorreclty called FLOPS or very roughly approximated as FLOPS via 2x scaling. Not having that would be a problem.

Having full coverage of all 'floating point operations' would be a nice as well, but it's not as clear cut as to what is / isn't a FLOP (there is a lot of room to debate) which is why I think the non-eager compiled version is good fit here, along with memory access, etc. You're more likely to care about counting every op in the context of what is actually being run in a given situation, where as GMAC is commonly used to compare model sizing / capacity and expected to be consistent regardless of the runtime.

**Comment by Chillee** [comment]
@albanD One issue is that I think it would be trickier to add the compiler-based version out of core.

> So for the eager version you really ignore everything that is not a matmul like conv/activations/normalization ? I feel like I'm missing something?

Well, I'm including "convs" in matmuls. But yeah, it's quite common to ignore non-matmul FLOPS, as they're generally quite negligible (and also involve lots of debate on how exactly to count them).

> For the eager version, can we make the design extensible such that if someone wants to add FLOPS it is easy?

Yeah, the eager version should be quite easily extensible.

**Comment by albanD** [comment]
> One issue is that I think it would be trickier to add the compiler-based version out of core.

Ho yes, I have no idea how you're going to do it so I trust you on this one.
One thing I would say is that, if that only requires a couple APIs within inductor for example to get the info you need, these will be most likely valuable APIs anyway (for people wanting to probe at what the generated code does) and so this is a good exercise to design them.

**Comment by Chillee** [comment]
PR for eager one is up here: https://github.com/pytorch/pytorch/pull/95751

**Comment by albanD** [comment]
Should we close this?

**Comment by Chillee** [comment]
No I still need to add the second version (the compiler-based flop counter).
